doc/io_streams.rdoc

```diff
@@ -0,0 +1,787 @@
+    f = File.new('t.tmp', 'w') # => #<File:t.tmp>
+    io = IO.new(f.fileno, 'w') # => #<IO:fd 4>
+
+    # Existing file; truncated to size zero.
+    File.write('t.tmp', text)
+    f = File.new('t.tmp', 'w') # => #<File:t.tmp>
+    f.size                     # => 0
+    io = IO.new(f.fileno, 'w') # => #<IO:fd 5>
+
+    # StringIO (not file).
+    sio = StringIO.new(text.dup, 'w') # => #<StringIO>
+
+  The stream is initially positioned at the beginning (which is also the end).
+  After each write, the stream is positioned at the end of the written data:
+
+    f.puts('First line')
+    f.flush
+    File.new('t.tmp').read # => "First line\n"
+    f.pos    # => 11
+    f.puts('Second line')
+    f.pos    # => 23
+    f.flush
+    File.new('t.tmp').read # => "First line\nSecond line\n"
+
+  You can use methods IO#rewind, IO#pos=, and IO#seek
+  to change to and read from any position (even past the stream's end):
+
+    f.pos = 0
+    f.write('FOO')
+    f.flush
+    File.new('t.tmp').read # => "FOOst line\nSecond line\n"
+
+    f.pos = 10
+    f.write('BAR')
+    f.flush
+    File.new('t.tmp').read # => "FOOst lineBARecond line\n"
+
+    f.pos = 24
+    f.write('BAZ')
+    f.flush
+    File.new('t.tmp').read # => "FOOst lineBARecond line\n\u0000\u0000BAZ"
+
+  Raises an exception if:
+
+  - Reading is attempted.
+  - The stream is to be a \File object,
+    but the given path exists and does not point to a writable file.
+  - The stream is to be an \IO object,
+    but the given file descriptor does not point to a writable file.
+
+- <tt>'a'</tt>: Write-only stream, positioned at the end:
+
+    # Non-existent file.
+    FileUtils.rm_f('t.tmp')
+    f = File.new('t.tmp', 'a')  # => #<File:t.tmp>
+    io = IO.new(f.fileno, 'a')  # => #<IO:fd 4>
+
+    # Existing file (not truncated).
+    File.write('t.tmp', text)
+    f = File.new('t.tmp', 'a')    # => #<File:t.tmp>
+    f.size                        # => 47
+    io = IO.new(f.fileno, 'a')    # => #<IO:fd 5>
+
+    # StringIO.
+    sio = StringIO.new(text,dup, 'a') # => #<StringIO>
+
+  Written data is always appended, regardless of the position
+  set by methods IO#rewind, IO#pos=, and IO#seek.
+
+    f.pos = 0
+    f.write('FOO')
+    f.flush
+    File.new('t.tmp').read
+    # => "First line\nSecond line\n\nFourth line\nFifth line\nFOO"
+
+    f.pos = 10
+    f.write('BAR')
+    f.flush
+    File.new('t.tmp').read
+    # => "First line\nSecond line\n\nFourth line\nFifth line\nFOOBAR"
+
+    f.pos = 800
+    f.write('BAZ')
+    f.flush
+    File.new('t.tmp').read
+    # => "First line\nSecond line\n\nFourth line\nFifth line\nFOOBARBAZ"
+
+  Raises an exception if:
+
+  - Reading is attempted.
+  - The stream is to be a \File object,
+    but the given path exists and does not point to a writable file.
+  - The stream is to be an \IO object,
+    but the given file descriptor does not point to a writable file.
+
+- <tt>'r+'</tt>: Read-write stream, positioned at the beginning:
```

Comment:
I've added a passage after the list of r, w, a, etc.  I agree that this is good to say, but am not sure where the best place is.  I didn't want to put it thrice.

Reply:
I think it would be useful in the blurb here, or somewhere right after the blurb. I think most readers would miss it if it's after the list.

Reply:
I'm not sure where "in the blurb here" is. Can you clarify?

Reply:
By "blurb" I mean the line following mode (which is "Read-write stream, positioned at the beginning" in this case).

Reply:
I'm considering an entirely different approach.  We might do well to doc this case-wise, depending on what the user wants to work with:

- Existing file (or file-based IO).
- New file.
- Existing IO (such as $stdin; not file-based).
- New IO (not file-based).
- Empty string.
- Non-empty string.

Then introduce each section with a table such as these,
to be followed by the usual details and examples.

```
=== Stream for Existing File

|------|----------|----------|----------|-----------|-----------|
| R/W  |          | Initial  |          | Initial   | Initial   |
| Mode |  Read    | Read Pos |  Write   | Write Pos | Truncate? |
|------|----------|----------|----------|-----------|-----------|
| 'r'  | Anywhere |    0     |   Error  |     -     |    No     |
| 'w'  |   Error  |    -     | Anywhere |     0     |    Yes    |
| 'a'  |   Error  |    -     | End only |    End    |    No     |
| 'r+' | Anywhere |    0     |   Error  |     -     |    No     |
| 'w+' | Anywhere |    0     | Anywhere |     0     |    Yes    |
| 'a+' | Anywhere |   End    | End only |    End    |    No     |
|------|----------|----------|----------|-----------|-----------|

=== Stream for File To Be Created

|------|----------|----------|----------|-----------|-----------|
| R/W  |          | Initial  |          | Initial   | Initial   |
| Mode |  Read    | Read Pos |  Write   | Write Pos | Truncate? |
|------|----------|----------|----------|-----------|-----------|
| 'r'  |   Error  |    -     |    -     |     -     |     -     |
| 'w'  | Anywhere |    -     | Anywhere |     0     |     -     |
| 'a'  |   Error  |    -     | End only |    End    |     -     |
| 'r+' |   Error  |    -     |    -     |     -     |     -     |
| 'w+' | Anywhere |    0     | Anywhere |     0     |     -     |
| 'a+' | Anywhere |   End    | End only |     -     |     -     |
|------|----------|----------|----------|-----------|-----------|
```



This would be a big change from the current PR, so I'd like to hear opinions before I invest a lot of work.

Also, I'm not positive that the two tables above are entirely correct, but that will need discussion only if we decide to have the tables.

Reply:
I would prefer to have class-specific documentation in the respective classes and this document to contain information common to these classes. The document becomes harder to navigate if it contains information that is specific to a certain class.

However, I am a big fan of the table. I think it makes it much easier to understand from a glance.

Reply:
I'm beginning to think that the read/write mode part of this (and maybe all of the mode) should be pushed out to the classes.  There's much more class dependency in the mode than I'd imagined.  I'll look into it.

Reply:
The thing is, it's not entirely class-based.  A file-based stream may be a File or an IO; so in IO, we'd need (at least) discussion of file-based and not.  Is a puzzlement!

Reply:
IMO it doesn't make sense to have file-specific mode information for IO since you can open all sorts of things in IO that are not files (e.g. stdin, stdout, pipes, sockets, etc.) where this mode information does not apply. I think this kind of information should belong only in the `File` class (and perhaps also `IO.sysopen`).

Reply:
I'm looking into it.

Reply:
I'm thinking to have a free-standing RDoc file titled "File-Based IO Streams" that can be linked from `File` (section "Access Modes") and `IO` (section "File-Based IO Object").

That new file would contain the tables above (corrections having been made), along with detailed sections corresponding to each table row.

`IO` would still need discussion of non-file-based instances.

Reply:
What about `IO.new(fd)`, where `fd` points to a file?

Reply:
Well the `fd` would have to have been opened via `File` or `IO.sysopen` right? It would depend on the mode that the file was opened with.

Reply:
Oh, dear.  So maybe the mode in `IO.new(fd, mode)` would be ignored?  Or would override?  Not sure I want to know.

Reply:
It seems to supplement the `mode` of the file descriptor.

`IO.new` raises an error if the two `mode` are not compatible:

```ruby
irb(main):001:0> fd = IO.sysopen("testing", "w")
=> 9
irb(main):002:0> IO.new(fd, "r")
(irb):2:in `initialize': Invalid argument (Errno::EINVAL)
```

And works if they're compatible:

```ruby
irb(main):003:0> fd = IO.sysopen("testing", "a+")
=> 10
irb(main):004:0> IO.new(fd, "r")
=> #<IO:fd 10>
```

Reply:
Done.  (Tables added.)

Reply:
"File Access Mode" moved to File intro.